pkg/artifactory/resource_artifactory_access_token.go

```diff
@@ -0,0 +1,330 @@
+				ValidateFunc: func(i interface{}, k string) (warnings []string, errors []error) {
+					v, ok := i.(string)
+					if !ok {
+						errors = append(errors, fmt.Errorf("expected type of %q to be string", k))
+						return warnings, errors
+					}
+
+					if _, err := time.Parse(time.RFC3339, v); err != nil {
+						errors = append(errors, fmt.Errorf("expected %q to be a valid RFC3339 date, got %q: %+v", k, i, err))
+					}
+
+					return warnings, errors
+				},
+			},
+			"refreshable": {
+				Type:     schema.TypeBool,
+				Optional: true,
+				Default:  false,
+				ForceNew: true,
+			},
+			"admin_token": {
+				Type:          schema.TypeSet,
+				ConflictsWith: []string{"groups"},
+				Optional:      true,
+				MaxItems:      1,
+				MinItems:      1,
+				ForceNew:      true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"instance_id": {
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: true,
+						},
+					},
+				},
+			},
+			"access_token": {
+				Type:      schema.TypeString,
+				Computed:  true,
+				Sensitive: true,
+			},
+			"refresh_token": {
+				Type:      schema.TypeString,
+				Computed:  true,
+				Sensitive: true,
+			},
+		},
+	}
+}
+
+func resourceAccessTokenCreate(d *schema.ResourceData, m interface{}) error {
+	client := m.(*ArtClient).ArtOld
+	grantType := "client_credentials" // client_credentials is the only supported type
+
+	tokenOptions := v1.AccessTokenOptions{}
+	resourceData := &ResourceData{d}
+
+	date, expiresIn, err := getDate(d)
+	if err != nil {
+		return err
+	}
+
+	tokenOptions.ExpiresIn = expiresIn
+	d.Set("end_date", date.Format(time.RFC3339))
+
+	refreshable := resourceData.Get("refreshable").(bool)
+	audience := d.Get("audience").(string)
+
+	tokenOptions.Audience = artifactory.String(audience)
+	tokenOptions.GrantType = &grantType
+	tokenOptions.Refreshable = artifactory.String(strconv.FormatBool(refreshable))
+	tokenOptions.Username = resourceData.getStringRef("username", false)
+
+	username := resourceData.Get("username").(string)
+	userExists, err := checkUserExists(client, username)
+
+	if !userExists && len(resourceData.Get("groups").([]interface{})) == 0 {
+		return fmt.Errorf("you must specify at least 1 group when creating a token for a non-existant user - %s, or correct the username", username)
+	}
+
+	err = unpackGroups(d, client, &tokenOptions)
+	if err != nil {
+		return err
+	}
+
+	err = unpackAdminToken(d, client, &tokenOptions)
+	if err != nil {
+		return err
+	}
+
+	AccessToken, _, err := client.V1.Security.CreateToken(context.Background(), &tokenOptions)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(strconv.Itoa(hashcode.String(*AccessToken.AccessToken)))
+	d.Set("access_token", *AccessToken.AccessToken)
+	if refreshable {
+		d.Set("refresh_token", *AccessToken.RefreshToken)
+	} else {
+		d.Set("refresh_token", "")
+	}
+
+	return nil
+}
+
+func resourceAccessTokenRead(d *schema.ResourceData, m interface{}) error {
```

Comment:
What's the purpose of this function? It's empty and thus should be removed.

Reply:
Hi @chb0github, the provider framework throws an error if Read function is not specified.

Reply:
Please add a comment to that effect then. It wasn't obvious to me.

Reply:
Okay. I've added a comment @chb0github